src/operator/tensor/dot-inl.h

```diff
@@ -443,46 +444,22 @@ struct DotCsrRspDnsByRowBlocks {
                                   const nnvm::dim_t nnr_r,
                                   const nnvm::dim_t num_rows,
                                   const nnvm::dim_t num_cols,
-                                  const nnvm::dim_t seg_len) {
+                                  const nnvm::dim_t seg_len,
+                                  const std::unordered_map<RType, nnvm::dim_t>* row_idx_map) {
     using nnvm::dim_t;
     const dim_t seg_start = i * seg_len;
     if (seg_start >= num_rows) return;
     const dim_t seg_end = std::min(seg_start + seg_len, num_rows);
     for (dim_t j = seg_start; j < seg_end; ++j) {
       if (indptr_l[j] == indptr_l[j+1]) continue;
       const dim_t offset_out = j * num_cols;
-      // Use binary search to find the lower_bound of val in row_idx array
-      const RType* first = row_idx_r;
-      const RType* last = row_idx_r + nnr_r;
-      const CType val = col_idx_l[indptr_l[j]];
-      const RType* it;
-      int count = last - first, step;
-      while (count > 0) {
-        it = first;
-        step = count / 2;
-        it += step;
-        if (*it < val) {
-          first = ++it;
-          count -= step + 1;
-        } else {
-          count = step;
-        }
-      }
-      const RType* row_idx_ptr = first;
-      // end of binary search
-      if (row_idx_ptr == row_idx_r+nnr_r || *row_idx_ptr > col_idx_l[indptr_l[j+1]-1]) continue;
-      for (IType k = indptr_l[j]; k < indptr_l[j+1] && row_idx_ptr != row_idx_r+nnr_r;) {
-        if (col_idx_l[k] == *row_idx_ptr) {
-          const dim_t offset_r = (row_idx_ptr - row_idx_r) * num_cols;
-          for (dim_t l = 0; l < num_cols; ++l) {
+      for (IType k = indptr_l[j]; k < indptr_l[j+1]; k++) {
+        for (dim_t l = 0; l < num_cols; ++l) {
+          if (row_idx_map->count(static_cast<RType>(k))) {
```

Comment:
`count` and `find` execute the same search twice. Use `find` here to get an iterator.